### PR TITLE
Joshua CAN brake light box code

### PR DESCRIPTION
**Pull request by joshua2020181**
SD card logging

### CODE REVIEW

**Review by rydzynskip — APPROVED** [review]
Looks great to me!

**Review by mccauleyma — APPROVED** [review]
Potentially overcomplicating the file naming process, but that's totally fine as long it doesn't cause issues. Schematic needs to be updated to show the SD Card module, and ofc overall this is pending actual testing within the CAN network

File: ArduinoCode/brake_light_box/brake_light_box.ino [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+  digitalWrite(CAN_SS_PIN, LOW); // enable CAN module
+
+  while (CAN_OK != CAN.begin(CAN_500KBPS, MCP_8MHz)) { //specify 8MHz crystal
+    Serial.println("CAN BUS init Failed");
+    delay(250);
+  }
+  Serial.println("CAN BUS Shield Init OK!");
+
+  digitalWrite(CAN_SS_PIN, HIGH); // disable CAN module
+  digitalWrite(SD_SS_PIN, LOW); // enable SD module
```
Comment (jamescd18) [inline ArduinoCode/brake_light_box/brake_light_box.ino:62]: this chunk could be a cool candidate for turning it into a separate `switchCANToSD()` function if you wanted. Totally optional.

File: ArduinoCode/brake_light_box/brake_light_box.ino [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+String errorBuffer = "";
+int bufferLength = 0;
+long lastLog = 0; // Time of last successful log to SD
+
+int nextFileName = 0; // Log file names in the format log-0.txt
+
+/**
+   Sets relavent pins to output or input, initializes serial for debugging, initializes CAN, SD card reader,
+   determines proper file name for future logs.
+*/
```
Comment (jamescd18) [inline ArduinoCode/brake_light_box/brake_light_box.ino:44]:
```suggestion
/**
 * Sets relavent pins to output or input, initializes serial for debugging, initializes CAN, SD card reader,
 * determines proper file name for future logs.
 */
```

File: ArduinoCode/brake_light_box/brake_light_box.ino [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+  digitalWrite(CAN_SS_PIN, LOW); // enable CAN module
+
+}
+
+bool SDWrite(String errors);
+void logError(String id, unsigned char error[8]);
+
+/**
+   Deals with incoming CAN messages, if an error is recieved, it is logged. Writes to SD card every 30 seconds.
+*/
```
Comment (jamescd18) [inline ArduinoCode/brake_light_box/brake_light_box.ino:100]:
```suggestion
/**
 * Deals with incoming CAN messages, if an error is recieved, it is logged. Writes to SD card every 30 seconds.
 */
```

File: ArduinoCode/brake_light_box/brake_light_box.ino [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    }
+  }
+}
+
+/**
+   Compiles errors, writes to SD every 5 errors
+
+   @param id the sender id of the error
+   @param error the errors to be logged
+*/
```
Comment (jamescd18) [inline ArduinoCode/brake_light_box/brake_light_box.ino:132]:
```suggestion
/**
 * Compiles errors, writes to SD every 5 errors
 *
 * @param id the sender id of the error
 * @param error the errors to be logged
 */
```

File: ArduinoCode/brake_light_box/brake_light_box.ino [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    }
+  }
+}
+
+/**
+   Logs errors to SD card.
+
+   @param errors the error(s) to be logged
+   @returns true if logging was successful, false if it failed
+*/
```
Comment (jamescd18) [inline ArduinoCode/brake_light_box/brake_light_box.ino:154]:
```suggestion
/**
 * Logs errors to SD card.
 *
 * @param errors the error(s) to be logged
 * @returns true if logging was successful, false if it failed
 */
```

**Review by jamescd18 — COMMENTED** [review]
Very well documented and I appreciate that since it makes everything so much easier to understand what it's doing. No particular issues per se, just optional areas where you could potentially refactor to make it shorter.

File: ArduinoCode/brake_light_box/brake_light_box.ino [hunk 6]
```diff
@@ -124,6 +122,22 @@ void loop() {
   }
 }
 
+/**
+ * Activates CAN module and deactivates SD module
```
Comment (jamescd18) [inline ArduinoCode/brake_light_box/brake_light_box.ino:126]:
```suggestion
 * Deactivates SD module and then activates CAN module
```

File: ArduinoCode/brake_light_box/brake_light_box.ino [hunk 7]
```diff
@@ -124,6 +122,22 @@ void loop() {
...
+ */
+void switchSDToCAN() {
+  digitalWrite(SD_SS_PIN, HIGH);
+  digitalWrite(CAN_SS_PIN, LOW);
+}
+
+/**
+ * Activates SD module and deactivates CAN module
```
Comment (jamescd18) [inline ArduinoCode/brake_light_box/brake_light_box.ino:134]:
```suggestion
 * Deactivates CAN module and then activates SD module
```

**Review by jamescd18 — CHANGES_REQUESTED** [review]
@rydzynskip can you confirm whether or not the suggestions I've made are correct? I just wanna be sure before pushing it all through

**Comment by rydzynskip** [comment]
The brake light schematic was updated so this was to assign the correct pins in the code.

**Review by jamescd18 — APPROVED** [review]
Looks good. Whoopsies on my part for the delay.

**Pull request merged**